paper-api/src/main/java/io/papermc/paper/world/damagesource/CombatTracker.java

```diff
@@ -0,0 +1,104 @@
+package io.papermc.paper.world.damagesource;
+
+import net.kyori.adventure.text.Component;
+import org.bukkit.entity.LivingEntity;
+import org.jetbrains.annotations.ApiStatus;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+import java.util.List;
+
+/**
+ * Represents entity's combat tracker
+ *
+ * @since 1.21.4
+ */
+@NullMarked
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface CombatTracker {
+
+    /**
+     * Gets the entity behind this combat tracker
+     *
+     * @return the entity behind this combat tracker
+     */
+    LivingEntity getEntity();
+
+    /**
+     * Gets the list of recorded combat entries.
+     * <br>
+     * Mutating the list is safe and won't have
+     * an effect on this entity's combat history.
+     *
+     * @return the list of combat entries
+     */
+    List<CombatEntry> getCombatEntries();
+
+    /**
+     * Sets the entity's combat history.
+     * <br>
+     * Note that overriding the entity's combat history won't
+     * have an effect on the entity's current or new combat state.
+     * Reset the current combat state and register new combat entries instead
+     * if you want the new history to have an effect on the combat state.
+     *
+     * @param combatEntries combat entries
+     * @see #resetCombatState()
+     * @see #addCombatEntry(CombatEntry)
+     */
+    void setCombatEntries(List<CombatEntry> combatEntries);
```

Comment:
What's the use of this?

Reply:
"Sneakily" setting prior combat history without overriding/resetting active combat state.
It's a neat shortcut for providing fake pre-made combat history scenarios and producing specific death messages.